### PR TITLE
[MHLO]: add BatchNormTraining and BatchNormGrad support in hlo-legalize-to-lhlo

### PR DESCRIPTION
**Pull request by lchang20**
It basically resolved lowering of some op with a result as a tuple of tensors.
There seem only BatchNormTraining and BatchNormGrad requiring this.
https://github.com/tensorflow/mlir-hlo/pull/18

### CODE REVIEW

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/hlo_legalize_to_lhlo.cc [hunk 1]
```diff
@@ -129,9 +131,22 @@ LogicalResult ConvertResults(Operation* op, SmallVectorImpl<Value>& results,
...
-    results.push_back(InsertDynamicAlloc(op->getLoc(), result.value(),
-                                         results_shape[result.index()],
+    results.push_back(InsertDynamicAlloc(op->getLoc(), resultType, op,
+                                         results_shape[index],
                                          &rewriter));
+    return success();
+  };
+
+  if (op->getNumResults() == 1 && op->getResult(0).getType().isa<TupleType>()) {
+    SmallVector<Type, 4> fattenedTypes;
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/hlo_legalize_to_lhlo.cc:141]:
```suggestion
    SmallVector<Type, 4> flattenedTypes;
```

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/hlo_legalize_to_lhlo.cc [hunk 2]
```diff
@@ -129,9 +131,22 @@ LogicalResult ConvertResults(Operation* op, SmallVectorImpl<Value>& results,
...
+    op->getResult(0).getType().cast<TupleType>().getFlattenedTypes(fattenedTypes);
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/hlo_legalize_to_lhlo.cc:142]:
```suggestion
    op->getResult(0).getType().cast<TupleType>().getFlattenedTypes(flattenedTypes);
```

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/hlo_legalize_to_lhlo.cc [hunk 3]
```diff
@@ -129,9 +131,22 @@ LogicalResult ConvertResults(Operation* op, SmallVectorImpl<Value>& results,
...
+    for (auto result : llvm::enumerate(fattenedTypes)) {
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/hlo_legalize_to_lhlo.cc:143]:
```suggestion
    for (auto result : llvm::enumerate(flattenedTypes)) {
```

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/hlo_legalize_to_lhlo.cc [hunk 4]
```diff
@@ -183,6 +198,37 @@ class HloToLhloOpConverter<mhlo::DotOp> : public BaseOpConversion<mhlo::DotOp> {
...
 
+template <typename HloOpTy>
+class HloTupleOutputToLhloOpConverter : public BaseOpConversion<HloOpTy> {
+public:
+  using BaseOpConversion<HloOpTy>::BaseOpConversion;
+  LogicalResult matchAndRewrite(
+    HloOpTy hloOp, ArrayRef<Value> operands,
+    ConversionPatternRewriter& rewriter) const final {
+    Operation* op = hloOp.getOperation();
+    SmallVector<Operation*, 4> allUsers;
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/hlo_legalize_to_lhlo.cc:209]:
```suggestion
    SmallVector<Operation*> allUsers;
```

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/hlo_legalize_to_lhlo.cc [hunk 5]
```diff
@@ -183,6 +198,37 @@ class HloToLhloOpConverter<mhlo::DotOp> : public BaseOpConversion<mhlo::DotOp> {
...
+    for (auto* user : op->getUsers()) {
+      allUsers.push_back(user);
+      if (!isa<mhlo::GetTupleElementOp>(user)) {
+        return failure();
+      }
+    }
+    SmallVector<Value, 4> buffer_args(operands.begin(), operands.end());
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/hlo_legalize_to_lhlo.cc:216]:
```suggestion
    SmallVector<Value> buffer_args(operands.begin(), operands.end());
```

File: tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/hlo_legalize_to_lhlo.cc [hunk 6]
```diff
@@ -183,6 +198,37 @@ class HloToLhloOpConverter<mhlo::DotOp> : public BaseOpConversion<mhlo::DotOp> {
...
+    if (failed(ConvertResults(op, buffer_args, rewriter))) return failure();
+    rewriter.create<mhlo::HloToLhloOp<HloOpTy>>(op->getLoc(), llvm::None,
+      buffer_args, op->getAttrs());
+    for (auto* user : allUsers) {
+      auto getElementOp = cast<mhlo::GetTupleElementOp>(user);
+      unsigned index = op->getNumOperands() + getElementOp.index();
+      getElementOp.getResult().replaceAllUsesWith(buffer_args[index]);
+      rewriter.eraseOp(user);
```
Comment (joker-eph) [inline tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/hlo_legalize_to_lhlo.cc:224]: This sequence isn't save: the replacement does not go through the `rewriter` and it won't be able to rollback or can gets its internal state confused. That can lead to crashes instead of diagnostics for example.

Let me try to rewrite this.

```suggestion
      rewriter.replaceOp(getElementOp, {buffer_args[index]});
```

**Review by joker-eph — APPROVED** [review]

**Comment by joker-eph** [comment]
Build is failing:
```
tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/hlo_legalize_to_lhlo.cc:210:50: error: non-virtual member function marked 'final' hides virtual member functions
      ConversionPatternRewriter& rewriter) const final {
```

**Comment by joker-eph** [comment]
```
third_party/llvm/llvm-project/mlir/include/mlir/IR/PatternMatch.h:1066:14: error: no matching member function for call to 'addDebugLabels'
    pattern->addDebugLabels(debugLabels);
    ~~~~~~~~~^~~~~~~~~~~~~~
third_party/llvm/llvm-project/mlir/include/mlir/IR/PatternMatch.h:1009:13: note: in instantiation of function template specialization 'mlir::RewritePatternSet::addImpl<mlir::mhlo::(anonymous namespace)::HloTupleOutputToLhloOpConverter<mlir::mhlo::BatchNormGradOp>, mlir::BufferizeTypeConverter &, mlir::MLIRContext *&>' requested here
        0, (addImpl<Ts>(/*debugLabels=*/llvm::None, arg, args...), 0)...};
            ^
```

**Comment by joker-eph** [comment]
```
third_party/llvm/llvm-project/mlir/include/mlir/IR/PatternMatch.h:280:9: error: cannot initialize object parameter of type 'const mlir::Pattern' with an expression of type 'mlir::mhlo::(anonymous namespace)::HloTupleOutputToLhloOpConverter<mlir::mhlo::BatchNormGradOp>'
    if (pattern->getDebugName().empty())
        ^~~~~~~~~
```

**Comment by lchang20** [comment]
> Build is failing:
> 
> ```
> tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/hlo_legalize_to_lhlo.cc:210:50: error: non-virtual member function marked 'final' hides virtual member functions
>       ConversionPatternRewriter& rewriter) const final {
> ```

I encountered this before when I accidentally used the master llvm.
It seemed that tensorflow might recently update the submoduled llvm version. 
Let me rebase it, and try to fix it.

**Comment by lchang20** [comment]
> > Build is failing:
> > ```
> > tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/hlo_legalize_to_lhlo.cc:210:50: error: non-virtual member function marked 'final' hides virtual member functions
> >       ConversionPatternRewriter& rewriter) const final {
> > ```
> 
> I encountered this before when I accidentally used the master llvm. It seemed that tensorflow might recently update the submoduled llvm version. Let me rebase it, and try to fix it.

Changed to the virtual function using adaptor.

**Review by joker-eph — APPROVED** [review]

**Comment by joker-eph** [comment]
Sorry this fell of my queue...

**Review by gbaned — APPROVED** [review]

**Comment by gbaned** [comment]
@lchang20 Any update on this PR? Please. Thanks!

**Comment by lchang20** [comment]
> @lchang20 Any update on this PR? Please. Thanks!

My bad, I thought it got merged. It should be fixed.
After tuple results becoming variadic, the fix became very simple. 
I just rollback and did minimal change on top of master.

**Comment by gbaned** [comment]
@lchang20 Can you please resolve conflicts? Thanks!

**Comment by lchang20** [comment]
> @lchang20 Can you please resolve conflicts? Thanks!

Done.

**Comment by gbaned** [comment]
Hi @joker-eph Can you please review this PR ? Thank you!

**Review by joker-eph — APPROVED** [review]

**Comment by joker-eph** [comment]
Oh wow that one really got missed for a long time...
